x-pack/legacy/plugins/file_upload/public/components/json_index_file_picker.js

```diff
@@ -21,89 +21,103 @@ export class JsonIndexFilePicker extends Component {
-          fileUploadError: i18n.translate(
-            'xpack.fileUpload.jsonIndexFilePicker.noFileNameError',
-            { defaultMessage: 'No file name provided' })
-        });
-        return;
-      }
-
-      // Check file type, assign default index name
-      const splitNameArr = file.name.split('.');
-      const fileType = splitNameArr.pop();
-      const types = ACCEPTABLE_FILETYPES.reduce((accu, type) => {
-        accu = accu ? `${accu}, ${type}` : type;
-        return accu;
-      }, '');
-      if (!ACCEPTABLE_FILETYPES.includes(fileType)) {
-        this.setState({
-          fileUploadError: (
-            <FormattedMessage
-              id="xpack.fileUpload.jsonIndexFilePicker.acceptableTypesError"
-              defaultMessage="File is not one of acceptable types: {types}"
-              values={{ types }}
-            />
-          )
-        });
+    if (fileArr.length === 0) { // Remove
+      return;
+    }
+    if (fileArr.length === 1) { // Parse & index file
+      const file = fileArr[0];
+      if (!this._validateFile(file)) {
         return;
       }
-      const initIndexName = splitNameArr[0];
-      setIndexName(initIndexName);
+      await this._parseFile(file);
+    }
+  };
 
-      // Check valid size
-      const { size } = file;
-      if (size > MAX_FILE_SIZE) {
-        this.setState({
-          fileUploadError: (
-            <FormattedMessage
-              id="xpack.fileUpload.jsonIndexFilePicker.acceptableFileSize"
-              defaultMessage="File size {fileSize} bytes exceeds max file size of {maxFileSize}"
-              values={{
-                fileSize: size,
-                maxFileSize: MAX_FILE_SIZE
-              }}
-            />
-          )
-        });
-        return;
-      }
+  _validateFile(file) {
+    const { resetFileAndIndexSettings, setIndexName } = this.props;
+    if (!file.name) {
+      this.setState({
+        fileUploadError: i18n.translate(
+          'xpack.fileUpload.jsonIndexFilePicker.noFileNameError',
+          { defaultMessage: 'No file name provided' })
+      });
+      return false;
+    }
 
-      // Parse file
-      this.setState({ fileParsingProgress: i18n.translate(
-        'xpack.fileUpload.jsonIndexFilePicker.parsingFile',
-        { defaultMessage: 'Parsing file...' })
+    // Check file type, assign default index name
+    const splitNameArr = file.name.split('.');
+    const fileType = splitNameArr.pop();
+    const types = ACCEPTABLE_FILETYPES.join(', ');
+    if (!ACCEPTABLE_FILETYPES.includes(fileType)) {
+      this.setState({
+        fileUploadError: (
+          <FormattedMessage
+            id="xpack.fileUpload.jsonIndexFilePicker.acceptableTypesError"
+            defaultMessage="File is not one of acceptable types: {types}"
+            values={{ types }}
+          />
+        )
+      });
+      resetFileAndIndexSettings();
+      return false;
+    }
+    const initIndexName = splitNameArr[0];
+    setIndexName(initIndexName);
```

Comment:
This seems like a side effect that I would not expect from calling a validate function. How about changing `_getIndexName` and throw if when file is invalid. That way, there will be no side effects in the function and `_fileHandler` can control all of the calls to `setIndexName`, `resetFileAndIndexSettings`, and setting state.

```
_getIndexName(file) {
  if (!file.name) {
    throw new Error(i18n.translate('xpack.fileUpload.jsonIndexFilePicker.noFileNameError', {
      defaultMessage: 'No file name provided'
    }));
  }

  const splitNameArr = file.name.split('.');
  const fileType = splitNameArr.pop();
  if (!ACCEPTABLE_FILETYPES.includes(fileType)) {
    throw new Error(i18n.translate('xpack.fileUpload.jsonIndexFilePicker.acceptableTypesError', {
      defaultMessage: 'File is not one of acceptable types: {types}'
      values: { 
        types: ACCEPTABLE_FILETYPES.join(', ')
      }
    }));
  }

  if (file.size > MAX_FILE_SIZE) {
    throw new Error(i18n.translate('xpack.fileUpload.jsonIndexFilePicker.acceptableFileSize', {
      defaultMessage: 'File size {fileSize} bytes exceeds max file size of {maxFileSize}'
      values: {
        fileSize: size,
        maxFileSize: MAX_FILE_SIZE
      }
    }));
  }

  return splitNameArr[0];
}
```

Then the caller could contain all of the logic of calling prop callbacks. This would make the logic easier to follow if all calls to prop callbacks are in same function and smaller logic subsets are abstracted behind function calls.

```
let initIndexName;
try {
  initIndexName = this._getIndexName(file);
} catch (error) {
  this.setState({
    fileUploadError: error.message
  });
  this.props.resetFileAndIndexSettings();
  return;
}

this.props.setIndexName(initIndexName);
```

Reply:
Nice. Sounds like a plan